Add new exclude_min and exclude_max attributes to InRangeValidator for numeric tool parameters.

gregvonkuster opened the pull request.

gregvonkuster committed:
Add a GreaterThanZeroValidator for numeric tool parameters.

nsoranzo commented:
For an integer `min="1"` should already do this, instead for a float it may be useful.

bgruening commented:
You can probably also use the `ExpressionValidator` with `int(value) > 0` but explicit is better than implicit, so I like it.

gregvonkuster commented:
@nsoranzo Are you referring to the InRangeValidator?  If so,I looked at it, but I'm implementing tools that draw 2D and 3D polygons using (x, y, z) coordinates, which must have edges > 0.0, but can have them less than 1.0, so setting a min value for the InRangeValidator will not work.  If you're referring to another validator, let me know as I did't see one that I could use, although I did miss the approach using the ExpressionValidator, so thanks @bgruening.  I'll use that one if this new validator is not appropriate.

nsoranzo commented:
@gregvonkuster I meant that for `IntegerToolParameter` and `FloatToolParameter` you can specify `min` and/or `max` attributes, but that just adds an `InRangeValidator`, so in practice yes. I also said that your validator may be of useful for floats for the exact reason you mentioned.

Maybe a more general approach would be to extend the `InRangeValidator` to let the user specify if the `min` should be strict (`value > min`) or not (`value >= min`), and similarly for the `max`.

gregvonkuster commented:
@nsoranzo Thanks for the feedback, I'll update my pull request with an enhancement to the InRangeValidator, eliminating the GreaterThanZeroValidator.

gregvonkuster committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into dev

gregvonkuster committed:
Enhance the InRangeValidator to enalbe the use of new strict_min and
strict_max attributes which, if set to true, will exclude the range end
points.  For example, if strict_min is true, min < value, whereas if
strict_min is false (the default), then min <= value.  Combinations of
strict_min and strict_max are allowed.

gregvonkuster commented:
This now contains an enhancement to the InRangeValodator as well as some pep-8 fixes.

bgruening commented on lib/galaxy/tools/parameters/validation.py:
```diff
@@ -103,20 +121,54 @@ class InRangeValidator( Validator ):
-        self.message = message or "Value must be between %s and %s" % ( self_min_str, self_max_str )
+        op1 = '>='
+        op2 = '<='
+        if self.strict_min and self.strict_max:
```
Hi @gregvonkuster this looks too complex. Isn't it simply

```
if self.strict_min:
    op1 = '>'
if self.strict_max:
    op2 = '<'
```

gregvonkuster replied:
@bgruening Maybe I'm not seeing it, but the various checks are needed only so that the one message line can use the defined operators based on the values of strict_min and strict_max.  I'm not seeing how it would be easier to set op1 and op2 to various combinations of <, <=, >, >= and still use a single message line.  Let me know what I'm missing and I'll update the pull request.  Thanks!

gregvonkuster replied:
@bgruening I just saw it - sorry!  My mind is tired.  Fix coming!

gregvonkuster committed:
Simplify the checks for setting strict_min and strict_max operators.

nsoranzo commented on lib/galaxy/tools/parameters/validation.py:
```diff
@@ -103,20 +121,54 @@ class InRangeValidator( Validator ):
+                raise ValueError( self.message )
+        elif not self.strict_min and not self.strict_max:
+            if not( self.min <= float( value ) <= self.max ):
+                raise ValueError( self.message )
```
What about the simpler:

```
if self.strict_min:
    if not self.min < float( value ):
        raise ValueError( self.message )
else:
    if not self.min <= float( value ):
        raise ValueError( self.message )
if self.strict_max:
    if not float( value ) < self.max:
        raise ValueError( self.message )
else:
    if not float( value ) <= self.max:
        raise ValueError( self.message )
```

gregvonkuster replied:
@nsoranzo thanks for the simplified validation logic.  I just committed it in another addition.

gregvonkuster committed:
Added @nsoranzo's simplified validation logic.

nsoranzo committed:
Simplify strict_min strict_max logic.

nsoranzo commented:
@gregvonkuster I created a PR on your repo to clarify what @bgruening and I mean.

gregvonkuster committed:
Merge pull request #1 from nsoranzo/for_greg

Simplify strict_min strict_max logic.

gregvonkuster commented:
@nsoranzo Sorry, I was pulled into something else for a while there - I've merged your pull request.  Thanks!

nsoranzo commented:
:+1: 

Thanks for merging! If you want to squash everything in a single commit, that's fine with me.

nsoranzo commented:
> If you want to squash everything in a single commit, that's fine with me.

Apparently this is considered bad practice, so forget about that.

jmchilton commented:
:+1: ... though I have a vague opinion that I would prefer `include_min` or `inclusive_min` language to the current `strict_min`.

Also - there is dissent on rebasing but that is irrelevant :).

nsoranzo commented:
@jmchilton: Or maybe `exclude_min`, which would be the exact boolean equivalent of `strict_min`.

@bgruening: Opinions? Merge as it is?

jmchilton commented:
Yeah - `exclude_min` is perfect. That language feels more natural and readable to me.

gregvonkuster committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into dev

gregvonkuster committed:
Merge branch 'dev' of https://github.com/gregvonkuster/galaxy into dev

gregvonkuster committed:
Change strict_mXX to exclude_mXX for InRangeValidator enhancement.

bgruening commented:
That both of you suggest completely different terms (`includes` vs. `excludes`) for the same feature is an indication that you can easily get it wrong, isn't it? What about `gt` or `lt`.

But I'm fine with `exclude_min` as well.

gregvonkuster commented:
@jmchilton @nsoranzo @bgruening  I made the requested change and renamed the pull request appropriately.  Thanks!

nsoranzo commented:
:+1:

Thanks @gregvonkuster for bearing with us!

gregvonkuster commented:
@nsoranzo Thanks for all of your help and the thumbs up!  :)
